### ISSUE TITLE
Trait refactor

### ISSUE THREAD

**Issue by str4d** [zkcrypto/group#6]
We've been working on refactoring the `group` crate traits over in https://github.com/zcash/librustzcash. See [here](https://github.com/zcash/librustzcash/blob/5ff85627448b407437b42fed4258386d2eda5ad5/group/src/lib.rs) for the current state of the refactor (as of 2020-05-28). The core of the current refactor is the following:
```
trait Group: GroupOps + GroupOps<<Self as Group>::Subgroup> {
    type Subgroup: PrimeGroup
    fn identity() -> Self::Subgroup;
}
trait PrimeGroup: Group {}
```
However, we can't easily convince `rustc` that `PrimeGroup` is only implemented on types for which `Group::Subgroup = Self`, and this then causes issues when using the group operations because `Self::Subgroup != Self::Subgroup::Subgroup`.

I'm now considering the following approach, where we define several base traits, and then separate top-level traits for prime-order and cofactor groups:
```rust
/// An element of a cryptographic group.
trait Group: GroupOps + ScalarMul<<Self as Group>::Scalar> {
    type Scalar: PrimeField;
    fn identity() -> Self;
    ...
}
/// Efficient representation of an elliptic curve point.
trait Curve: Group + GroupOps<<Self as Curve>::AffineRepr> { // Mixed addition
    type AffineRepr;
    fn to_affine() -> Self::AffineRepr;
    ...
}
/// A group element with a defined encoding.
trait GroupEncoding {
    type Repr: Default + AsRef<[u8]> + AsMut<[u8]>;
    fn from_bytes(&Self::Repr) -> CtOption<Self>;
    ...
}
/// Affine representation of an elliptic curve point that has a defined uncompressed encoding.
trait UncompressedEncoding {
    type Uncompressed: Default + AsRef<[u8]> + AsMut<[u8]>;
    fn from_uncompressed(&Self::Uncompressed) -> CtOption<Self>;
    ...
}

/// An element of a prime-order cryptographic group.
trait PrimeGroup: Group + GroupEncoding {}
trait PrimeCurve: Curve<AffineRepr = <Self as PrimeCurve>::Affine> + PrimeGroup {
    type Affine: PrimeCurveAffine<Curve = Self, Scalar = Self::Scalar>
        // Scalar multiplication with mixed addition
        + Mul<Self::Scalar, Output = Self>;
}
pub trait PrimeCurveAffine:
    GroupEncoding
    // Scalar multiplication with mixed addition
    + Mul<<Self as PrimeCurveAffine>::Scalar, Output = <Self as PrimeCurveAffine>::Curve>
{
    type Scalar: PrimeField;
    type Curve: PrimeCurve<Affine = Self, Scalar = Self::Scalar>;
    fn identity() -> Self;
    fn to_curve(&self) -> Self::Curve;
    ...
}

trait CofactorGroup:
    Group
    + GroupEncoding
    // Using the type system to enforce Group + Subgroup -> Group
    + GroupOps<<Self as CofactorGroup>::Subgroup>
{
    /// The large prime-order subgroup in which cryptographic operations are performed.
    /// If `Self` also implements `PrimeGroup`, then `Self::Subgroup` may be `Self`.
    type Subgroup: PrimeGroup<Scalar = Self::Scalar> + Into<Self>;
    fn mul_by_cofactor(&self) -> Self::Subgroup;
    fn into_subgroup(self) -> CtOption<Self::Subgroup>;
    ...
}
trait CofactorCurve: Curve<AffineRepr = <Self as CofactorCurve>::Affine> + CofactorGroup {
    type Affine: CofactorCurveAffine<Curve = Self, Scalar = Self::Scalar>
        // Scalar multiplication with mixed addition
        + Mul<Self::Scalar, Output = Self>;
}
trait CofactorCurveAffine:
    GroupEncoding
    // Scalar multiplication with mixed addition
    + Mul<<Self as CofactorCurveAffine>::Scalar, Output = <Self as CofactorCurveAffine>::Curve>
{
    type Scalar: PrimeField;
    type Curve: CofactorCurve<Affine = Self, Scalar = Self::Scalar>;
    fn identity() -> Self;
    fn to_curve(&self) -> Self::Curve;
    ...
}
```

I see several advantages for this kind of API:
- `PrimeGroup` does not have a `Subgroup` associated type, bounding any recursion.
- Protocol implementors can either restrict their implementations to only work with `PrimeGroup`s, or can explicitly choose to support `CofactorGroup`s (and then explicitly handle the subgroup conversion with e.g. `CofactorGroup::mul_by_cofactor`, which would be a no-op for `PrimeGroup`s).
- Protocol implementors can choose to specialise to elliptic curves if they want to leverage an affine representation and mixed addition in their protocol for efficiency, or they can ignore those traits and stick with the simpler group-focused traits.

I'd like to hear thoughts on this; in particular, what pain points this might cause for current users of the `group` crate vs the current (elliptic-curve-specific) traits.

**Comment by huitseeker** [comment]
I think this is a really good proposal, given motivated implementers.

The main disadvantage with this API is that implementers have to duplicate a bit of tedious work between the prime group and the cofactor group.

Would the operations you're looking to implement work better if you:
- required the `Group` trait to come equipped with a (canonical) surjection into the subgroup pi,
- asked for a representant function hoisting back the subgroup to the group — say repr,
- asked implementers to define this implementation such that pi is a left inverse for repr¹.

The you would be able to define the group operations generically, and I'd expect Rust would be very much able to elide the use of `std::convert::identity` (encouraged by doc examples) that would occur in the prime subgroup cases.

¹: If you really want to armor this, you could ask for implementations of `proptest::Arbitrary` for each and unit-test the property.

**Comment by str4d** [comment]
> The main disadvantage with this API is that implementers have to duplicate a bit of tedious work between the prime group and the cofactor group.

For a cofactor group with a prime subgroup, this is true. We cannot simply implement `CofactorGroup { Subgroup: Self }` , because while generically `Self` and `Self::Subgroup` will be distinct types, concretely this requires an `impl PrimeGroup for Self` which is a type error; thus we need a separate subgroup newtype, and associated pass-through logic for implementing `Group` etc on the subgroup newtype.

For a prime group, `CofactorGroup` can be implemented trivially with `CofactorGroup { Subgroup: Self }` and trivial implementations of the API.

> I wonder if the operations you're looking to implement would work if you:
> - required the Group trait to come equipped with a (canonical) surjection into the subgroup pi,

This is provided by `CofactorGroup::mul_by_cofactor` and `CofactorGroup::into_subgroup`. Which of these should be used is context-specific, but in e.g. element decoding we know we want to use `into_subgroup` inside `impl GroupEncoding for Subgroup`.

> - asked for a representant function hoisting back the subgroup to the group — say repr,

We constrain `CofactorGroup::Subgroup: Into<Self>` which provides this.

> - asked implementers to define this implementation such that pi is a left inverse for repr¹.

This should hopefully be obvious, but we can make this explicit somewhere in the documentation if necessary.

> The I suspect you would be able to define the group operations generically

I don't think this is possible _completely_ generically without having the subgroup type explicitly encode `Self` as an associated type (and thus know that it has an inner type implementing `Group` that it can defer to). This would impose cofactor-relevant implementation details on the prime group's API (something I am attempting to avoid in this design).

However, if we are in the implementation context and know that e.g. the subgroup is implemented as a newtype wrapper around the general group element type, then the pass-through operations on the subgroup type are well-defined, and could be provided as a `group` crate macro.

**Comment by str4d** [comment]
I've updated the OP with the current state of the design (only a few small changes), and have implemented it in https://github.com/zcash/librustzcash/pull/241. Please experiment with the implementation (particularly after that PR is merged) and give feedback here!

**Comment by str4d** [comment]
The trait refactor was released in 0.7.0 and merged here in #7.
